### ISSUE TITLE
Validation error when attempting to upload a document with no attributes assigned - BPMS I-15

### ISSUE THREAD

**Issue by willdwyer** [willdwyer/bcbsmaissuestracker#253]
Hello,

Please see steps below.
1. Go to Content Library
2. Go to any random folder
3. Click New, Document (to upload)
4. Fill out criteria to upload a document
5. (assign attributes window populates below)
6. User attempts to click submit with no attributes assigned

![screen shot 2015-09-02 at 4 08 42 pm](https://cloud.githubusercontent.com/assets/13453008/9643014/572bd816-518d-11e5-91a8-ffaa83dbccf8.png)

Basically, we would like that users MUST assign certain attributes when uploading a document, otherwise they receive a validation error. Please let us know if this possible.

The specific attributes we would like to prompt a validation error have yet to be determined, but will be specified soon.  

@jddecoste @g4gaurang

**Comment by michparrish** [comment]
Hi Will,

Current functionality does not include a concept of required attributes. This would be an enhancement request.
Please let me know if you would like us to submit this request for consideration.

Thank you,
Michelle

**Comment by willdwyer** [comment]
Hi Michelle,

I was out all of last week, so am just getting back to this item now.

Can you provide a estimate level of effort, and timing for completion?

Thanks,

**Comment by michparrish** [comment]
Will - I can create the enhancement request and it will go into the queue for an upcoming sprint to assess consideration of the change (b/c this may be considered a system-wide change and may not be considered for enhancement) and LOE.

**Comment by willdwyer** [comment]
Hi Michelle,

Just want to confirm - this is in the queue for sprint 8, correct? (9/8 - 9/29)

Do you have an update on this?  Please let me know.

**Comment by michparrish** [comment]
Hi Will - no, this has not been queued up for a sprint yet. I will update the log when it is scheduled for one and then provide an update once the sprint is complete.

**Comment by willdwyer** [comment]
Hello,

We need to add to this enhancement request.  See screenshot below.  It was noticed earlier that if a user tries to assign attributes with documents of 2 different types, they are unable to.  We will need a user to be able to assign attributes for 2 document types at once, as with plans there will be both a word document and PDF associated with them.  Without this override, it will take double the time to upload and assign attributes.

![screen shot 2015-09-28 at 1 50 50 pm](https://cloud.githubusercontent.com/assets/13453008/10144063/7ec6ba38-65e9-11e5-9e3d-25dbaa8206f7.png)

Please provide an LOE and timing for this as well as the enhancement request referenced above.

**Comment by willdwyer** [comment]
Hello,

We also encountered a defect tying in with this enhancement request.  Per above screenshot, steps below...

(1)if a user goes to upload and assign attributes for 2 different document types
(2)receives error message
(3)deletes one of the 2 document types
(4)an error pops up saying "please contact system administrator"
(5)user is then prevented from assigning attributes to the single document remaining

As mentioned above, we are requesting an enhancement for a user to (1)receive a validation error if they try to upload a document with no attributes assigned to it, as well as (2)be able to edit attributes for documents of different document types.

![screen shot 2015-09-29 at 10 40 39 am](https://cloud.githubusercontent.com/assets/13453008/10168052/0b404a60-6699-11e5-84f9-1dc12892fc3e.png)
![screen shot 2015-09-29 at 10 42 41 am](https://cloud.githubusercontent.com/assets/13453008/10168066/16d98026-6699-11e5-937f-dad6be0fd9ad.png)
![screen shot 2015-09-29 at 10 42 52 am](https://cloud.githubusercontent.com/assets/13453008/10168069/19006810-6699-11e5-93c5-1e06f32c3e92.png)

**Comment by michparrish** [comment]
I am going to open a new item for the additional request - we cannot roll them into one another so far apart, as #15 is already included in this sprint.

**Comment by michparrish** [comment]
@willdwyer Question on the error you're receiving - you shouldn't be able to upload different doc types to the same folder at all. SBC PDF Finals go into a different folder than SBC Word Finals, so you have to have 2 different uploads anyway.
Please confirm.

**Comment by michparrish** [comment]
I have sent a note to QA to ask about the system error - I still received this error when I was uploading 2 of the same doc types but it works fine if you click through it. I was able to assign attributes successfully as well.

**Comment by willdwyer** [comment]
Hi Michelle,

That is partially correct.  The user should not be able to upload different doc types to the same folder.  However, a user does have the ability to upload multiple doc types at once.  For example, in the above screenshot, if the user were to upload those 2 different doc types (.pdf and .doc) in the same instance, then there would be no issue as long as the user doesn't try to assign attributes to them.  The upload would just prompt the .doc file to be moved to the Word folder and the .pdf file to be moved to the PDF folder.

I know you're going to be at BCBS tomorrow, so I can walk you through this in person (probably easier).  I showed Steve A this earlier.

**Comment by michparrish** [comment]
Hi Will - thank you, I'll test again. I got the system error when I tried this but I'll try again. I will not be in the office tomorrow but will be on Friday.

**Comment by michparrish** [comment]
Retested and I see there is still a system error when uploading different doc types, but you're still successful when clicking through it and resubmitting. I also see that the Attribute assignment window is now closed for use when different doc types are listed. I have opened the new enhancement request for that capability, which is separate from this one and will be tracked separately.
I am working with QA re: the system administration error message.

**Comment by willdwyer** [comment]
OK great.  Thanks, Michelle.  As mentioned, please let us know the LOE and timing estimate for this when you can.

**Comment by willdwyer** [comment]
Hi Michelle,

Any update on the LOE and timing for this enhancement?

**Comment by michparrish** [comment]
This is under consideration in this sprint and we should have a response posted by 11/4.

**Comment by willdwyer** [comment]
Hi Michelle,

Any update on this?

**Comment by michparrish** [comment]
Hi Will, I have a request out to get the status of all of the Defects and Enhancement Requests for LOE that were included in this sprint. I hope to get this today and update all of the items here in Github.

**Comment by TimHeasley** [comment]
Sprint was not able to accommodate estimates, bugs were first priority.

**Comment by michparrish** [comment]
@willdwyer 
The product team has come back with a recommendation for functionality that is already in the system and can just be turned on.
The functionality is called 'Recommended Attributes', whereas I denote whichever attributes you want to force a user to update to be 'recommended'. When the user uploads a document or uses Tools > Edit Attributes, these specific attributes are listed at the top of the list in their own 'box' under the heading of 'Recommended Attributes'. When a user tries to Submit an upload or Submit attribute updates without changing these values from the default 'N/A' value, a popup renders stating there are still Recommended Attributes with their default values, do they want to continue. This prompts the user to hit Cancel and then update the attributes.

We cannot actually make any of the Attributes required b/c, in a sense, they actually already are. Each attribute requires a default value, which we've set to 'N/A' for all of them since the value cannot be blank. We cannot remove the Default functionality, the only thing we could do is change the default value to another value but this does not help the 'required' need.
Notifying the user that they have not made changes from the value of 'N/A' to a more applicable value in the list is an option we can provide as a solution and is an easy one (very low LOE) since the functionality is already in place.

Please let me know if you would like to discuss or, if you're good with this explanation, what your team decision is. Below are screen shots from pclient5 and you are more than welcome to go in there yourself and test it out. Try uploading a doc into Plan Masters or select a plan design and go to Tools > Edit Attributes.

Thank you!
[Required Attributes.docx](https://github.com/g4gaurang/bcbsmaissuestracker/files/30313/Required.Attributes.docx)

**Comment by willdwyer** [comment]
Hi Michelle,

I will have to discuss this internally, and test/experiment with it, but personally... I like the looks of it.  I will  do some testing with it in PCLIENT5, and get back to you on this.  I will leave this assigned to me.

**Comment by willdwyer** [comment]
Michelle, just a few questions...
1. Is there a limit to how many attributes can be assigned as "Recommended"?
2. Are there any attributes we are unable to assign as "Recommended"?
3. Can documents to-be-uploaded of any type have the Recommended attribute designation to them?

**Comment by michparrish** [comment]
1. Nope - no limit
2. You can assign any of the UDAs (User Defined Attributes) but not the SDAs (System Defined Attributes) - see list attached
   ![uda list](https://cloud.githubusercontent.com/assets/13471811/11047748/1d810b36-8702-11e5-9d1b-b7821ace282e.JPG)
3. Yes, this would be across Content Library, whenever you're uploading any document anywhere.

**Comment by willdwyer** [comment]
Thanks, Michelle.  We will review this internally.  We have an internal meeting scheduled for 11/12, and will have a response to you by then.  I will leave this assigned to me for now.

**Comment by michparrish** [comment]
sounds good, thank you!
And just to give a number, the LOE would be about an hour. It's very easy to set up and you can essentially test them all at the same time.

**Comment by willdwyer** [comment]
OK, thanks.  I added an LOE-1hr tag to this.

**Comment by willdwyer** [comment]
Hi Michelle,

The below is the list of all attributes we will need to be assigned as 'recommended'.

Plan Year
Product Group Type
Market Segment
Financial Arrangement
Product Type
Product Identifier
Effective Date
Anniversary Date
Plan Design Type

Please add these as recommended, and we will do some testing in PCLIENT5.

**Comment by michparrish** [comment]
All set! Let me know how the testing goes!

**Comment by willdwyer** [comment]
Hi Michelle,

I apologize, but Product Identifier was not meant to be in the list of Recommended Attributes.  It should be in the optional list.

Please remove, and we will proceed with retesting.

**Comment by michparrish** [comment]
all set

**Comment by willdwyer** [comment]
Looks good, please move to production.

**Comment by michparrish** [comment]
all set. Updated the following attributes to be 'Recommended' in Production:

Plan Year
 Product Group Type
 Market Segment
 Financial Arrangement
 Product Type
 Effective Date
 Anniversary Date
 Plan Design Type

Please test out and close out item if satisfied. Thanks!

**Comment by willdwyer** [comment]
Hi Michelle,

One thing i just discovered while doing some testing in production...

Can the recommended attribute list also be applied to the Bulk Edit Attribute functionality?  See screenshot below.

![screen shot 2015-11-16 at 2 11 50 pm](https://cloud.githubusercontent.com/assets/13453008/11192006/98de143a-8c6c-11e5-9460-b19c5e9f6f1e.png)

**Comment by michparrish** [comment]
Hi Will,
Unfortunately that is the one place we cannot invoke this functionality since essentially it's a spreadsheet for updates. But it will be under Tools > Edit Attributes and when you upload a document.

**Comment by michparrish** [comment]
Will - sorry, I misspoke a bit and was thinking you were talking about the Output Attributes bulk editing functionality. That said, with any instance of bulk editing, we're not able to enforce this because users can edit documents of many types, the Recommended Attribute functionality was deemed impossible to enforce. This specific functionality was put in place for users to select specific document types to note 'recommended attributes' for. I have placed it on "All" doc types, so whenever someone uploads a new doc of any kind, they have this 'recommended' popup come up. 
I can open up an item for Dev to look into adding this functionality to Bulk Editing when the denotation of 'All' is selected for Doc Types - please let me know if this is an enhancement request you'd like me to open.
Thanks!

**Comment by willdwyer** [comment]
Michelle,

So, would this be an entirely new enhancement request?  (assigning recommended attributes for bulk editing)?

**Comment by michparrish** [comment]
right - so the original request was to have required attributes at the time of a doc upload. Our proposal for that item was to use the 'Recommended Attributes' functionality that we've now implemented, which also pertains to Tools > Edit Attributes of one document.
Since this does not apply to bulk editing of attributes, we would have to open a new request for this new request to be looked into, as it pertains to different functionality in the system.

**Comment by willdwyer** [comment]
OK, that is fine.  Please open up a new enhancement request for providing recommended attributes on the bulk edit screen.  Note that we will need an LOE and timing estimate before deciding whether or not to proceed.

**Comment by michparrish** [comment]
Will do - do we want to close this item out as resolved since we handled the original issue and then open a new enhancement request in GitHub to track separately? I can open the item if you want.

**Comment by willdwyer** [comment]
Yes, this item can be closed.
